plugins/chat/app/services/services.md

```diff
@@ -0,0 +1,153 @@
+* [MessageMover](https://github.com/discourse/discourse/blob/f34b4f67b1f15c51ecba9ad73b4e39515cf9e988/plugins/chat/lib/message_mover.rb)
+* [SlackCompatibility](https://github.com/discourse/discourse/blob/f34b4f67b1f15c51ecba9ad73b4e39515cf9e988/plugins/chat/lib/slack_compatibility.rb)
+* [ChatChannelFetcher](https://github.com/discourse/discourse/blob/f34b4f67b1f15c51ecba9ad73b4e39515cf9e988/plugins/chat/lib/chat_channel_fetcher.rb)
+* [ChatMessageReactor](https://github.com/discourse/discourse/blob/f34b4f67b1f15c51ecba9ad73b4e39515cf9e988/plugins/chat/lib/chat_message_reactor.rb)
+* [ChatChannelMembershipManager](https://github.com/discourse/discourse/blob/f34b4f67b1f15c51ecba9ad73b4e39515cf9e988/plugins/chat/lib/chat_channel_membership_manager.rb)
+
+These services all do different things. Some are purely for data fetching (e.g.
+ChatChannelFetcher), others are used for MessageBus publishing (e.g.
+ChatMessagePublisher), while still more handle complicated processes that
+include doing things like this (taking ChatMessageCreator as an example):
+
+* Validating input parameters
+* Raising errors based on failed validation, access control, and other issues
+* Querying the database
+* Validating current database record state
+* Changing database record state
+* Triggering DiscourseEvent events
+* Calling other services
+* Enqueueing jobs
+* Checking permissions with Guardian
+* Logging, both to STDOUT and to things like StaffActionLogger
+* Interacting with MessageBus
+
+When writing services, we also need to think about how we handle failure and
+success states, especially for the Controllers that are most often calling them
+(though of course services can and do call other services). With that in mind we
+need to handle each different function of services differently to make it easier
+for the developer.
+
+#### Errors
+
+We are doing this a few different ways right now, here are some examples:
+
+* Raising errors, either custom ones or things like `Discourse::InvalidParameters`
+  as a form of control flow
+* Calling Guardian `ensure_X!` variants which raise an error if the condition is
+  not met
+* Returning a true/false value and message based on the result
+* Catching raised errors within the service and expecting the caller to check
+  for them
+
+Most errors need to be shown to the user as well (though not all). I think we
+already have a better solution for this -- the `HasErrors` module,
+which behaves similarly to ActiveModel::Errors. Using this, we can:
+
+* Add arbitrary error messages
+* Add all errors from ActiveRecord models when calling their own .validate
+  Methods
+
+Sometimes we need to stop the world when we find one of these errors, other
+times we can just collect them and show them all at the end. The base class for
+service objects should support both. We can define many ease-of-use methods in
```

Comment:
Im mixed on this, errors that don't stop are not errors IMO. They are some kind of specific events in some services which could be treated in the result object:

```
if not_processed_count > 0
  context.audit_log << "#{not_processed_count} messages have not been processed"
end
```

And then have some logic in controller rendering if `audit_log` is not empty.

Failure should be a strong failure IMO.